### PR TITLE
Download Yarn Berry with Corepack rather than Yarn 1.x

### PR DESCRIPTION
**Pull request by basil**
Decrease the maintenance burden in this repository by replacing our custom Ant (!) based code to bootstrap Yarn Berry using Yarn 1.x with Corepack as implemented in https://github.com/eirslett/frontend-maven-plugin/pull/1157.

### Testing done

Introduced stylelint and eslint failures in both CI and local builds. In both cases, verified that the error message was readable: in local builds, on the CLI; and in CI builds, in the Jenkins UI.

For the local builds, tested without path modifications on my local machine with Node LTS installed as well as with path modifications inside of a Docker container with no existing Node/NPM installation, thus exercising both supported paths as documented in `CONTRIBUTING.md`.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

```[tasklist]
### Submitter checklist
- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.
```

### Desired reviewers

@mention

Before the changes are marked as `ready-for-merge`:

```[tasklist]
### Maintainer checklist
- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).
```

### CODE REVIEW

File: .gitignore
Comment (basil) [inline .gitignore:68]: See the comments to `.yarnrc.yml` below for explanation.

File: .gitignore
Comment (basil) [inline .gitignore:70]: Previously automatically generated on the fly by our custom Ant-based logic. This Ant-based logic is now deleted, and instead this file is checked into Git. This removes a layer of indirection and makes the code simpler, since the file that is used is now static rather than dynamically generated.

File: .gitignore
Comment (basil) [inline .gitignore:81]: Using a more specific path to `jsbundles`, since there is only a single such directory.

File: .gitignore
Comment (basil) [inline .gitignore:84]: Using NPM is not supported, and having code in the source tree to handle an unsupported configuration by silently ignoring it (rather than failing fast) violates the principle of least surprise.

File: .prettierignore
Comment (basil) [inline .prettierignore:10]: No longer automatically generated, so we can adhere to Prettier formatting rules now.

File: .yarnrc.yml
Comment (basil) [inline .yarnrc.yml:2]: As recommended in https://github.com/eirslett/frontend-maven-plugin/blob/4a501a10677716a0e6676b603e3b93bc6bf319f1/frontend-maven-plugin/src/it/corepack-provided-integration/.yarnrc.yml. Note that without the `nodeLinker` line https://github.com/eirslett/frontend-maven-plugin/pull/1157 does not actually work, instead complaining about `corepack` not being listed in `package.json`. I filed for another day any latent curiosity about why this might be the case (probably a deficiency  in `frontend-maven-plugin` documentation at the very least).

File: CONTRIBUTING.md
Comment (basil) [inline CONTRIBUTING.md:56]: Clarifying that we only support Node.js 20.x LTS.

I changed the first occurence of "`yarn`" to "Yarn" because this is a reference to the Yarn project, not the `yarn` binary. Even though we later use "`yarn` binary" in the same sentence, this is not inconsistent because the first usage refers to the project and the second usage refers to the binary.

Elsewhere in the sentence we replace "PATH" with "path" for consistency with the beginning of the sentence and the rest of the document. While both "`PATH`" (as a particular variable name) or "path" (referring to the general concept of a search path) would be correct here, this document generally uses the latter, so be consistent with usage elsewhere in this sentence and in this document.

File: CONTRIBUTING.md
Comment (basil) [inline CONTRIBUTING.md:61]: Adapting to the changes in `pom.xml` below.

File: war/.gitignore
Comment (basil) [inline war/.gitignore:1]: Unneeded now that the corresponding entries are in the root directory's `.gitignore` file.

File: pom.xml
Comment (basil) [inline pom.xml:103]: Yarn 1.x is now no longer needed to bootstrap Yarn Berry, since we now use Corepack for this purpose.

File: pom.xml
Comment (basil) [inline pom.xml:106]: This duplicate information is no longer needed, as this information is consumed from `package.json` via Corepack.

File: CONTRIBUTING.md
Comment (scherler) [reply CONTRIBUTING.md:56]: > Clarifying that we only support Node.js 20.x LTS.

I am not sure if that statement is true. If that is what we want we should then change the `package.json` to add 

```
"engines": {
  "node": ">=20.0.0
},
```

**Comment by scherler** [comment]
Tried to build but it is failing 
image

File: .yarnrc.yml
Comment (scherler) [reply .yarnrc.yml:2]: why not using 'pnp' https://yarnpkg.com/migration/pnp

Comment (basil) [reply .yarnrc.yml:2]: I see no reason why using PNP would solve the problem statement described in this PR description.

**Comment by basil** [comment]
> Tried to build but it is failing image

That seems like evidence that you have failed to build Jenkins core, but it does not seem like evidence that there is a general issue with this PR, as the **Testing Done** section is thorough, including both local testing and CI testing. I am inclined to believe there is something wrong on your side, but that this should not block us from proceeding with this PR.

**Comment by timja** [comment]
My build failed with:

```
[INFO] Downloading https://registry.npmjs.org/corepack/-/corepack-provided.tgz to /Users/timja/.m2/repository/com/github/eirslett/corepack/provided/corepack-provided.tar.gz
[INFO] No proxies configured
[INFO] No proxy was configured, downloading directly
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Jenkins main module 2.478-SNAPSHOT:
[INFO]
[INFO] Jenkins main module ................................ FAILURE [  3.534 s]
[INFO] Jenkins BOM ........................................ SKIPPED
[INFO] Internal SPI for WebSocket ......................... SKIPPED
[INFO] Jetty 12 (EE 9) implementation for WebSocket ....... SKIPPED
[INFO] Jenkins cli ........................................ SKIPPED
[INFO] Jenkins core ....................................... SKIPPED
[INFO] Jenkins war ........................................ SKIPPED
[INFO] Tests for Jenkins core ............................. SKIPPED
[INFO] Jenkins coverage ................................... SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  5.905 s
[INFO] Finished at: 2024-09-23T20:31:22+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.eirslett:frontend-maven-plugin:1.15.1:install-node-and-corepack (install node and corepack) on project jenkins-parent: Could not download corepack: Got error code 404 from the server. -> [Help 1]
[ERROR]
```

Steps I followed were:

```bash
gh pr checkout 9772
mvn clean install -P quick-build
```

Env:
```
jenkins [🌱 corepack][$?][☕ v17][⬢ v20.9.0][🐪 v5.34.0][🐍 v3.12.5][⏱ 7s]
❯ pwd
/Users/timja/code/jenkins/jenkins
```

Apple M2, on Sequoia

File: CONTRIBUTING.md
Comment (timja) [reply CONTRIBUTING.md:56]: with this change do we need this sentence? (as nothing will automatically update this.

**Comment by timja** [comment]
I managed to fix it by running `rm -rf node/`

and then the next downloaded version was setup correctly.

I wasn't able to reproduce it by trying to revert to my previous state:

```
rm -rf node/
git checkout master
mvn clean install -P quick-build
git checkout -
mvn clean install -P quick-build
```

File: .yarnrc.yml
Comment (timja) [inline .yarnrc.yml:1]: What's the reasoning for disabling this? it seems to be enabled by default.

(not a blocker)

File: CONTRIBUTING.md
Comment (basil) [reply CONTRIBUTING.md]: > with this change do we need this sentence?

I suppose not, since Yarn [apparently enforces the version specified by `engines`](https://chrisshepherd.me/posts/enforcing-nodejs-npm-yarn-versions)—unlike NPM. I haven't tested this, though. I don't feel strongly about discussing the issue further, so I have simply removed the change in commit c6a69c8908. With the change now removed, there ought not to be a reason to discuss the issue further, either.

File: .yarnrc.yml
Comment (basil) [reply .yarnrc.yml:1]: This was copied from the example here: https://github.com/eirslett/frontend-maven-plugin/blob/4a501a10677716a0e6676b603e3b93bc6bf319f1/frontend-maven-plugin/src/it/corepack-provided-integration/.yarnrc.yml

There is no comment there explaining the reason, but if I had to guess, it would be to follow the general philosophy of `frontend-maven-plugin`, as explained in its `README`:

> Node/corepack and any package managers will only be "installed" locally to your project. It will not be installed globally on the whole system (and it will not interfere with any Node/corepack installations already present).

The design seems to be to remain completely isolated from the system installation of Node/Corepack, and I suppose the developer was trying to adhere to that philosophy here by avoiding the sharing of any caches with the system installation.

**Comment by basil** [comment]
> I managed to fix it

OK, your original error message is definitely related to this PR, and I am glad you were able to solve it by cleaning the working directory. I had tested with a clean working directory as well. Just to be sure, I ran a build successfully in a Docker container without any existing Maven cache, and it worked correctly there, so I think this PR has no issues on a clean environment. So I think we will just need to make sure that developers who have an existing checkout run `git clean -fxd` or similar before trying to build after this PR. If I am wrong, though, and there is some persistent failure with this PR that is not present before this PR, I would investigate that.

**Review by timja — APPROVED** [review]
Thanks!

**Comment by daniel-beck** [comment]
It's not a new phenomenon that moving around JS related files breaks the local build. The same happened due to #9718 when switching between "before" and "after" branches, and the same happened a few years ago. Perhaps this just is the first time it happens when moving forward rather than backward.

**Comment by basil** [comment]
> It's not a new phenomenon that moving around JS related files breaks the local build with unclean contents. The same happened due to #9718 when switching between "before" and "after" branches, and the same happened a few years ago. Perhaps this just is the first time it happens when moving forward rather than backward.

Indeed, this is yet another ancient build bug. I will try to fix `mvn clean` to work properly with `frontend-maven-plugin` in a PR after this one.

**Review by scherler — APPROVED** [review]

**Comment by timja** [comment]
/label ready-for-merge

----

This PR is now ready for merge, after the pending security release, we will merge it if there's no negative feedback.

Thanks!

**Pull request merged**
